src/cpu/testers/traffic_gen/gups_gen.cc

```diff
@@ -0,0 +1,414 @@
+        requestPool.pop();
+
+        if (requestPool.size() < reqQueueSize) {
+            doGenerate = true;
+        }
+    }
+
+    if (!nextSendEvent.scheduled() && (doGenerate ||
+        (!port.blocked() && !requestPool.empty()))) {
+        schedule(nextSendEvent, nextCycle());
+    }
+
+    return;
+}
+
+
+void
+GUPSGen::GenPort::sendTimingPacket(PacketPtr pkt)
+{
+    panic_if(_blocked, "Should never try to send if blocked MemSide!");
+
+    // If we can't send the packet across the port, store it for later.
+    if (!sendTimingReq(pkt)) {
+        DPRINTF(GUPSGen, "GenPort::sendTimingPacket: Packet blocked\n");
+        blockedPacket = pkt;
+        _blocked = true;
+    }
+    else {
+        DPRINTF(GUPSGen, "GenPort::sendTimingPacket: Packet sent\n");
+    }
+
+}
+
+void
+GUPSGen::GenPort::sendFunctionalPacket(PacketPtr pkt)
+{
+    sendFunctional(pkt);
+}
+
+void
+GUPSGen::GenPort::recvReqRetry()
+{
+    // We should have a blocked packet if this function is called.
+    DPRINTF(GUPSGen, "GenPort::recvReqRetry: Received a retry.\n");
+
+    assert(_blocked && (blockedPacket != nullptr));
+    // Try to resend it. It's possible that it fails again.
+    _blocked = false;
+    sendTimingPacket(blockedPacket);
+    if (!_blocked){
+        blockedPacket = nullptr;
+    }
+
+    owner->wakeUp();
+}
+
+bool
+GUPSGen::GenPort::recvTimingResp(PacketPtr pkt)
+{
+    owner->handleResponse(pkt);
+    return true;
+}
+
+GUPSGen::GUPSGenStat::GUPSGenStat(GUPSGen* parent):
+    Stats::Group(parent),
+    ADD_STAT(totalTicks,
+        "Total number of ticks for this generator to generate requests"),
+    ADD_STAT(totalUpdates,
+        "Total number of updates the generator made in the memory"),
+    ADD_STAT(GUPS,
+        "Number of updates per second"),
+    ADD_STAT(totalReads,
+        "Total number of read requests"),
+    ADD_STAT(totalBytesRead,
+        "Total number of bytes read"),
+    ADD_STAT(avgReadBW,
+        "Average read bandwidth received from memory"),
+    ADD_STAT(totalReadLat,
+        "Total latency of read requests."),
+    ADD_STAT(avgReadLat,
+        "Average latency for read requests"),
+    ADD_STAT(totalWrites,
+        "Total number of read requests"),
+    ADD_STAT(totalBytesWritten,
+        "Total number of bytes read"),
+    ADD_STAT(avgWriteBW,
+        "Average read bandwidth received from memory"),
+    ADD_STAT(totalWriteLat,
+        "Total latency of write requests."),
+    ADD_STAT(avgWriteLat,
+        "Average latency for read requests")
+{}
+
+void
+GUPSGen::GUPSGenStat::regStats(){
+    GUPS.precision(8);
+    avgReadBW.precision(2);
+    avgReadLat.precision(2);
+    avgWriteBW.precision(2);
+    avgWriteLat.precision(2);
+
+    GUPS = (totalUpdates * 1e3) / totalTicks;
+
+    avgReadBW = (totalBytesRead * 1e12) / totalTicks;
+    avgReadLat = (totalReadLat * 1e-3) / totalReads;
+
+    avgWriteBW = (totalBytesWritten * 1e12) / totalTicks;
+    avgWriteLat = (totalWriteLat * 1e-3) / totalWrites;
```

Comment:
This can be done in the constructor.

BTW, I *suggest* (just a suggestion) not having any formulas. I think that formulas should be done in post-processing.